_posts/2018-05-23-getting-foreman-search-results-into-your-puppet-manifest.md

```diff
@@ -0,0 +1,98 @@
+This page is an update of the [previous post written by Ohad Levy](https://theforeman.org/2012/01/getting-foreman-search-results-into.html)
+
+<!--more-->
+
+## Why ?
+
+When you start Puppet, in general you start simply. And so innocently, each node is more or less alone in the world. And Puppet, in this way, can already help you so much.
+
+But quickly, you understand that in some situations, it is interesting to know information from other nodes. And this happens typically about monitoring, backups or identity management. For instance, you add a new node into your infrastructure. And your monitoring server takes into account this new server by adding the configuration relating to its monitoring. 
+
+In a standard way, in the Puppet galaxy there exists the option `storeconfigs` of Puppet Server and linked [PuppetDB](https://puppet.com/docs/puppetdb/latest). A PuppetDB service records every node's catalogs, facts and reports. And the immediate usage is exported resources (a double “at” sign `@@` in manifests). 
+
+But, if you don't have a PuppetDB and you are already using Foreman as external node classifier (aka ENC), you can also make fun stuff.
+
+## An use case out of standard usages.
+
+Use cases about monitoring or backups are standard. Lets try to build another example where we can use the Foreman search engine.
+
+Imagine that you are managing nodes with a Puppet Server and they are in `hostgroups` like `secure_base/role`. Where `secure_base` permits to apply all common settings to all your nodes in a secured environment. And `role` permits to apply settings dedicated to the role of a node. Now, lets say you have the constraint to not use a DNS server. And your customer would like to call hosts in a human way, so with names and not with IPs. I already met this use case. Fortunately, servers were using only one IP address.
+
+So you have to manage your `/etc/hosts` with a configuration management tool or you become crazy. 
+
+If each Puppet node execute the following code :
+
+```
+  $query = foreman({foreman_user => 'apiuser',
+                    foreman_pass => 'here_a_password',
+                    item         => 'hosts',
+                    search       => 'hostgroup_fullname ~ secure_base',
+  })
+
+  $nodes = $query['results']
+
+```
+
+In `$nodes`, you got all the needed informations about other nodes. You can check my statement by adding the following piece of code :
+
+
+```
+  $nodes.each | $_node | {
+    $_thiscert = $_node['certname']
+    $_thisip = $_node['ip']
+    notify {  "node ${_thiscert}" : message => "host ${_thisip}"}
+  }
+
+```
+
+Let's use `$nodes` in a template with a `file` resource to manage `/etc/hosts` and you got the trick.
+
+## Unknow function foreman()
+
+If you ran the code above, you probably got an error message like `Evaluation Error: Unknown function: 'foreman'`. By default, the function `foreman()` is not available.
+
+Several possibilities to get the function into your environement :
```

Comment:
typo: environment

Any reason you're not advising to just install `theforeman-foreman` in your environment?

Reply:
In general, i prefer to limit the amount of unknown code introduced into my environment.